ambry-store/src/main/java/com.github.ambry.store/PersistentIndex.java

```diff
@@ -306,22 +314,71 @@ private void recover(MessageStoreRecovery recovery) throws StoreException, IOExc
    * @throws IOException if an end offset could not be set due to I/O error.
    */
   private void setEndOffsets() throws IOException {
-    for (Offset segmentStartOffset : indexes.keySet()) {
-      Offset nextIndexSegmentStartOffset = indexes.higherKey(segmentStartOffset);
+    for (Offset segmentStartOffset : indexesRef.get().keySet()) {
+      Offset nextIndexSegmentStartOffset = indexesRef.get().higherKey(segmentStartOffset);
       if (nextIndexSegmentStartOffset == null || !segmentStartOffset.getName()
           .equals(nextIndexSegmentStartOffset.getName())) {
         // this is the last index segment for the log segment it refers to.
-        IndexSegment indexSegment = indexes.get(segmentStartOffset);
+        IndexSegment indexSegment = indexesRef.get().get(segmentStartOffset);
         LogSegment logSegment = log.getSegment(segmentStartOffset.getName());
         logSegment.setEndOffset(indexSegment.getEndOffset().getOffset());
       }
     }
   }
 
   /**
+   * @return the map of {@link Offset} to {@link IndexSegment} instances.
+   */
+  ConcurrentSkipListMap<Offset, IndexSegment> getIndexSegments() {
+    return indexesRef.get();
+  }
+
+  /**
+   * Atomically adds {@code segmentFilesToAdd} to and removes {@code segmentsToRemove} from the map of {@link Offset} to
+   * {@link IndexSegment} instances.
+   * @param segmentFilesToAdd the backing files of the {@link IndexSegment} instances to add.
+   * @param segmentsToRemove the start {@link Offset} of {@link IndexSegment} instances to remove.
+   * @throws IllegalArgumentException if any {@link IndexSegment} that needs to be added or removed has an offset higher
+   * than that in the journal.
+   * @throws StoreException if an {@link IndexSegment} instance cannot be created for the provided files to add.
+   */
+  void changeIndexSegments(List<File> segmentFilesToAdd, Set<Offset> segmentsToRemove) throws StoreException {
+    TreeMap<Offset, IndexSegment> segmentsToAdd = new TreeMap<>();
+    Offset journalFirstOffset = journal.getFirstOffset();
+    for (File indexSegmentFile : segmentFilesToAdd) {
+      IndexSegment indexSegment = new IndexSegment(indexSegmentFile, true, factory, config, metrics, journal);
+      if (indexSegment.getEndOffset().compareTo(journalFirstOffset) > 0) {
+        throw new IllegalArgumentException("One of the index segments has an end offset " + indexSegment.getEndOffset()
+            + " that is higher than the first offset in the journal " + journalFirstOffset);
+      }
+      segmentsToAdd.put(indexSegment.getStartOffset(), indexSegment);
+    }
+
+    indexesChangeLock.lock();
+    try {
+      ConcurrentSkipListMap<Offset, IndexSegment> oldIndexSegments = indexesRef.get();
+      ConcurrentSkipListMap<Offset, IndexSegment> newIndexSegments = new ConcurrentSkipListMap<>();
+      for (Map.Entry<Offset, IndexSegment> entry : oldIndexSegments.entrySet()) {
+        if (!segmentsToRemove.contains(entry.getKey())) {
```

Comment:
This is scenario I think the lock is required for. Consider two threads, one of them executing `addToIndex()` and the other `changeIndexSegments()`.

The one executing `changeIndexSegments()` has just executed the line copying all the existing index segments over to what will be the new map (`newIndexSegments = new ConcurrentSkipListMap<>(oldIndexSegments)`) but hasn't switched the atomic ref to the new map (i.e. `indexesRef.set(newIndexSegments)` has not been executed). The one executing `addToIndex()` decides that a rollover is required and adds an `IndexSegment` - note that this all to the "old" index segment map since the ref hasn't been switched yet. Now, when the thread executing `changeIndexSegments()` does the `set()` operation, the newly added `IndexSegment` is lost.

Let me know if you see any holes in the reasoning or still think a lock isn't required.

Took your suggestions on the refactoring. Makes the code clearer, thanks !